File: lib/core/utils/semana.dart
// ============================================================
//  Identificador de semana ISO-8601 (puro, sin Flutter)
//
//  Punto único para derivar el periodo semanal de los leaderboards
//  (`leaderboards/{periodo}`). Espejo de `functions/src/semana.ts`.
// ============================================================

/// Devuelve el identificador de semana ISO-8601 de [fecha] como `YYYY-Www`
/// (p. ej. `2026-W24`).
///
/// La semana empieza en **lunes** y la semana 1 es la que contiene el primer
/// jueves del año; el año del identificador es el del jueves de esa semana (por
/// eso los últimos días de diciembre pueden caer en la semana 1 del año
/// siguiente, y viceversa). Se calcula en UTC para que el periodo no dependa de
/// la zona horaria del dispositivo.
String idSemanaIso(DateTime fecha) {
  // Normaliza a medianoche UTC; el jueves de la semana define año y número.
  final d = DateTime.utc(fecha.year, fecha.month, fecha.day);
  final jueves = d.add(Duration(days: 4 - d.weekday)); // weekday: 1=lun..7=dom
  final inicioAnio = DateTime.utc(jueves.year, 1, 1);
  final semana = 1 + (jueves.difference(inicioAnio).inDays ~/ 7);
  return '${jueves.year}-W${semana.toString().padLeft(2, '0')}';
}

File: functions/src/semana.ts
// ─────────────────────────────────────────────────────────────────────────────
// Identificador de semana ISO-8601 (`YYYY-Www`). Espejo de
// lib/core/utils/semana.dart. Define el periodo de los leaderboards
// (`leaderboards/{periodo}`). Se calcula en UTC.
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Devuelve el identificador de semana ISO-8601 de [fecha] como `YYYY-Www`.
 * Semana en lunes; la semana 1 contiene el primer jueves del año; el año es el
 * del jueves de esa semana.
 */
/**
 * Día calendario en UTC como `YYYY-MM-DD`. Base de la racha del bono diario
 * (Fase 10b): consistente entre cliente y servidor por usar UTC.
 */
export function idDiaUtc(fecha: Date): string {
  return fecha.toISOString().slice(0, 10);
}

export function idSemanaIso(fecha: Date): string {
  const d = new Date(
    Date.UTC(fecha.getUTCFullYear(), fecha.getUTCMonth(), fecha.getUTCDate()),
  );
  // getUTCDay: 0=domingo..6=sábado → normalizar a 1=lunes..7=domingo.
  const diaSemana = d.getUTCDay() === 0 ? 7 : d.getUTCDay();
  // Avanzar al jueves de esta semana.
  d.setUTCDate(d.getUTCDate() + 4 - diaSemana);
  const inicioAnio = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
  const semana = Math.ceil(
    ((d.getTime() - inicioAnio.getTime()) / 86400000 + 1) / 7,
  );
  return `${d.getUTCFullYear()}-W${String(semana).padStart(2, '0')}`;
}
